In [1]:
pip install transformers

Note: you may need to restart the kernel to use updated packages.


## Low Level API

In [2]:
from transformers import AutoTokenizer, AutoModel

#load tokenizer and model
tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")
model = AutoModel.from_pretrained('bert-base-uncased')

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

BertModel LOAD REPORT from: bert-base-uncased
Key                                        | Status     |  | 
-------------------------------------------+------------+--+-
cls.predictions.transform.dense.weight     | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED |  | 
cls.predictions.bias                       | UNEXPECTED |  | 
cls.seq_relationship.bias                  | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.weight | UNEXPECTED |  | 
cls.predictions.transform.dense.bias       | UNEXPECTED |  | 
cls.seq_relationship.weight                | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


we just loaded
* pretrained **Bert tokenizer**
* pretrained **Bert model**

these were pretrained on massive amount of text datasets

In [3]:
#tokenize text
text = 'I love machine learning'

inputs = tokenizer(text, return_tensors='pt')
print(inputs)

{'input_ids': tensor([[ 101, 1045, 2293, 3698, 4083,  102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1]])}


* input_ids = numbers representing tokens
* attention_mask = (0: padding, 1: real token)

In [5]:
#pass through the model
outputs = model(**inputs)
print(outputs.last_hidden_state.shape) #(batch_size, sequence_length, hidden_size)

torch.Size([1, 6, 768])


Meaning:
* 1 sentence
* 6 tokens
* each token → 768-dimensional vector

Unlike normal embeddings: 
* word -> fixed vector

BERT does:
* word -> context based embeddings

Example:
"bank" in:
- river bank
- bank account

Different meanings → different embeddings.

## High level API

In [6]:
from transformers import pipeline

classifier = pipeline('sentiment-analysis')
result = classifier('I love this course')
print(result)

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision 714eb0f.
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/104 [00:00<?, ?it/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

[{'label': 'POSITIVE', 'score': 0.999883770942688}]


What just happened internally was:

        Text
        → Tokenizer
        → Model (BERT)
        → Classification head
        → Output

So there are two ways to use Hugging Face transformers they are as follows:
* High level API: pipeline()  This is the easy method
* Low level API: AutoTokenizer + AutoModel  This is more powerful